README.md

```diff
@@ -1,10 +1,148 @@
+
+This package currently provides support for Firebase out of the box but allows developers to configure it to work with any backend/database.
+
+## 🚀 Quick Setup (Next.js Example)
+
+### Install
+
+```shell
+npx create-next-app crossmint-solana-auth --example "https://github.com/crossmint/solana-auth/packages/examples/nextjs-starter"
+
+cd crossmint-solana-auth && yarn install
+yarn dev
+```
+
+## 🛠 Setup (React + Firebase)
+
+## Install
+
+```sh
+yarn add @crossmint/solana-auth-base@beta  \
+         @crossmint/ solana-auth-react-ui@beta
+```
+
+Define `signIn` and `signOut`. Here we use firebase utils and `signInWithCustomToken`.
+
+```js
+// utils/firebaseClient.ts
+export async function signIn(data: Record<string, string>) {
+  const userCredential = await signInWithCustomToken(auth, data.token);
+  return userCredential;
+}
+
+export const signOut = () => {
+  return auth.signOut();
+};
+```
+
+Wrap your app in `SolanaAuthProvider`. Make sure to pass it a domain, your auth fucntions, and your serverless endpoints.
+
+```javascript
+// _app.tsx
+import { AppProps } from "next/app";
+import { FC } from "react";
+import { SolanaAuthProvider } from "../src";
+import { signIn, signOut } from "../utils/firebaseClient";
+
+import { SolanaAuthProvider } from "@crossmint/solana-auth-react-ui";
+
+require("../styles/globals.css");
+
+const App: FC<AppProps> = ({ Component, pageProps }) => {
+  return (
+    <SolanaAuthProvider
+      domain="example.xyz"
+      requestUrl="/api/solana-auth/getauthchallenge"
+      callbackUrl="/api/solana-auth/completeauthchallenge"
+      onAuthCallback={signIn}
+      signOut={signOut}
+    >
+      <Component {...pageProps} />
+    </SolanaAuthProvider>
+  );
+};
+
+export default App;
+```
+
+Use the `SolanaFirebaseAuth` component to render the signin button in your app.
+
+```js
+import { SolanaAuthButtonFirebase } from "@crossmint/solana-auth-react-ui";
+import { auth } from "../utils/firebaseClient";
+const Index = () => {
+  return (
+    <div>
+      <head>
+        <title>Sign in with Solana</title>
+        <meta name="description" content="Sign in with Solana" />
+        <link rel="icon" href="/favicon.ico" />
+      </head>
+
+      <main>
+        <SolanaAuthButtonFirebase auth={auth} />
+      </main>
+    </div>
+  );
+};
+
+export default Index;
+```
+
+## Use the serverless functions.
+
+```js
+import { FirebaseAdapter, SolanaAuth } from "@crossmint/solana-auth-base";
+import firebaseApp from "../../../utils/firebase";
+
+export default SolanaAuth({ adapter: FirebaseAdapter(firebaseApp) });
+```
+
+This object exposes `getAuthChallenge` and `completeAuthChallenge`. These functions take a request and response like vercel or firebase functions and return `Promise<void>`.
+
+## Client Usage
+
+export default Index;
+
+````
+
+# Custom Configuration
+
+## Defining an adapter
+
```

Comment:
Database adapters are needed in order to store transient log-in attempt information. We provide a built-in adapter for Firebase Firestore, or you can implement your own.